### ISSUE TITLE
Tests time out non-deterministically

### ISSUE THREAD

**Issue by keyboardDrummer** [boogie-org/boogie#909]
Different tests seems to fail non-deterministically:

- CodeExpr0.bpl: https://github.com/boogie-org/boogie/actions/runs/9973516141/job/27559010448?pr=908
- test2/B.bpl: https://github.com/boogie-org/boogie/actions/runs/9940226812/job/27456456838

**Comment by keyboardDrummer** [comment]
Problem started with this PR: https://github.com/boogie-org/boogie/pull/899

**Comment by keyboardDrummer** [comment]
These testruns did not have the problem: 

```
Wed, 03 Jul 2024 14:32:37 GMT: https://github.com/boogie-org/boogie/actions/runs/9779824193/job/26999970017
Current runner version: '2.317.0'
Operating System
  Ubuntu
  [2](https://github.com/boogie-org/boogie/actions/runs/9779824193/job/26999970017#step:1:2)2.04.4
  LTS
Runner Image
  Image: ubuntu-22.0[4](https://github.com/boogie-org/boogie/actions/runs/9779824193/job/26999970017#step:1:4)
  Version: 20240[6](https://github.com/boogie-org/boogie/actions/runs/9779824193/job/26999970017#step:1:7)24.1.0
```

These are some of the first test runs suffering from it:

```
Thu, 04 Jul 2024 15:17:49: https://github.com/boogie-org/boogie/actions/runs/9796660534/job/27051593251
```
Tue, 09 Jul 2024 14:00:44: https://github.com/boogie-org/boogie/actions/runs/9858481941/job/27219967914

**Comment by keyboardDrummer** [comment]
One thing that's a shame is that the timeout comes from lit, not from the Boogie process. It would be good to configure timeouts at the Boogie level that are shorter than the lit timeout, so that when the timeout occurs, the Boogie process can give some feedback as to where it occurred.

**Comment by keyboardDrummer** [comment]
Desperate move but maybe upgrading dotnet will help: https://github.com/boogie-org/boogie/issues/909

**Comment by atomb** [comment]
> The issue seems to come from the image version upgrade from 20240624.1.0 to 20240630.1.0, since all runs with the problem are in 20240630.1.0 and none are on 20240624.1.0.

The diff between those Ubuntu image versions is [here](https://github.com/actions/runner-images/commit/7863994db86a8163ca12950be0e1fda4b6f85491), but it unfortunately gives me no ideas as to what might be going wrong. I can't imagine how any of those changes would have impacted anything.

One interesting data point is that the normal runtime for the tests that are hanging is extremely short. Verification time for each is a few milliseconds, and the Z3 resource count is usually under 1000, and always under 3000.

My best hypothesis is that this is a deadlock. Maybe the changes to the underlying system are somehow enough to make the non-determinism of concurrency more likely to go wrong?

It could be worth seeing what happens if we try reverting concurrency-related PRs and seeing whether that makes any difference (not that we'd want to revert them permanently, because I think they all fix important things, but it might help us understand what's going on). The ones I see in the recent past include #841, #854, #888, and _maybe_ #898?

**Comment by keyboardDrummer** [comment]
> My best hypothesis is that this is a deadlock.

I agree. I increased the timeout by a huge amount and the timeouts still occurred.
